control_apps/src/control_mngr.rs

```diff
@@ -0,0 +1,292 @@
+        return self.mode;
+    }
+
+    pub fn set_target(&mut self, target_altitude: f32) {
+        // set new target altitude
+        // Set a new target altitude to converge toward (in meters)
+        if target_altitude > CTRL_ALTITUDE_FLOOR {
+            // target must be above the minimum allowed altitude
+            self.target_altitude = target_altitude;
+            info!("New target altitude: {:}m", self.target_altitude);
+        } else {
+            warn!(
+                "Not allowed to set target altitude below {:}! Ignoring...",
+                CTRL_ALTITUDE_FLOOR
+            );
+        }
+    }
+
+    pub fn start_control(&mut self) {
+        // enable altitude control+transition to idle
+        let previous_mode = self.get_mode();
+        if previous_mode == ControlMode::Safe || previous_mode == ControlMode::Idle {
+            info!("Enabling altitude control system!");
+            // unlock the valves if they are locked to force the next step!
+            self.valve_vent.unlock();
+            self.valve_dump.unlock();
+            // execute mode transition
+            self.mode = ControlMode::Idle;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "Not allowed to enable control from {:} mode! Ignoring...",
+                previous_mode
+            );
+        }
+    }
+
+    pub fn power_on_self_test(&mut self) {
+        // turn on and test devices to look for errors
+        info!("Starting Power-On Self Test...");
+        // hard-coded wait to check
+        thread::sleep(Duration::from_millis(POST_DELAY_MS));
+        info!("POST complete!");
+        // when POST is complete, transition from idle to safe
+        self.safe()
+    }
+
+    pub fn update(
+        &mut self,
+        gps_altitude: f32, // instantaneous altitude in meters from GPS
+        ballast_mass: f32, // ballast mass remining in kg
+        delta_t: f32,      // time since last update
+    ) {
+        // execute control algorithm and update vent and dump valve PWM values
+
+        // check abort criteria
+        if !self.abort_if_too_low(gps_altitude) && !self.abort_if_out_of_ballast(ballast_mass) {
+            // update error
+            let last_error = self.altitude_error;
+            let error = gps_altitude - self.target_altitude;
+
+            // change mode if applicable, run control with gain switching
+            if error.abs() >= CTRL_ERROR_DEADZONE {
+                debug!(
+                    "Altitude error {:}m greater than deadzone threshold {:}m",
+                    error, CTRL_ERROR_DEADZONE
+                );
+                if error > 0.0 {
+                    // lower altitude for error to converge to zero
+                    debug!(
+                        "Altitude {:}m is higher than target {:}m, vent gas",
+                        gps_altitude, self.target_altitude
+                    );
```

Comment:
I think this is a bit untidy to be interacting with `self.valve_vent` here right after doing so in `self.vent()`.  Couldn't this be bumped up?

Also see comments on `vent()`

Reply:
I guess I saw `self.vent()` and `self.dump()` as just the action to switch states, but the action to update the PWM would be kicked off by the scope of the `self.ipdate()` function.

A reason why I don't have the system "select" a valve and the later just update pwm for whichever valve is selected is because I couldn't figure out a clean way to do it and still respect the mutable/borrowing rules.

I don't see why not to add a `valve_vent.update_pwm()` call to the `self.vent()` method (or dump respectively)... I guess up until now I only saw the `.vent()` method being used to initiate a transition. But reading through it would in fact be called every update. Hmm

Reply:
Check the latest update. I migrated the pwm updates into the state methods.
I still think there's a better way to do this but I'm not sure about it.